### ISSUE TITLE
What's the current status/capabilities of the bridge?

### ISSUE THREAD

**Issue by MarioAlmelaUMH** [Robotics010/carla_autoware_bridge#8]
Dear @Robotics010

Thank you very much for this new initiative. My name is Mario and I work as a researcher at Universidad Miguel Hernandez (Spain). We have recently started to work with Carla and Autoware, and have tried different existing bridges. Our expertise is on V2X communications and would like to consider these tools for a more accurate modeling of the environment and driving. We may consider the adoption of your bridge (now or in the future) if it aligns with our needs. To this end, we would like to ask you a few questions about its status and roadmap and technical capabilities:

1) Current status. You mentioned that it is still a Work in Progress package. Can you please let us know what you mean? What are the basic features currently supported by the bridge and do you know what are its main limitations? 

2) Roadmap. Since the bridge has been recently published, we understand that there will be further development to enhance the bridge and overcome the potential limitations it may have in its first version. Can you please confirm that and let us know if there is a list of features already in the roadmap? We are thinking about the implementation of the radar (as soon as it is available in Autoware) and the migration to ROS humble (understanding the limitations of Carla). Are these two in the development plan?

3) Scenarios. Carla allows generating very complex scenarios. Is the bridge capable of supporting any Carla scenarios or is there any non-supported Carla feature? Like, for example, generating multiple vehicles/pedestrians or using custom maps.

4) Multi-sensor capabilities. We’ve seen in this discussion (https://github.com/orgs/autowarefoundation/discussions/3391) that Autoware currently launches multiple lidars and that shouldn’t be a problem for the bridge. Is it possible to use different sensor_kits (or even custom ones), or does the bridge only support Autoware’s standard sensor_kit? Can the bridge support multiple lidars and multiple cameras installed in the ego vehicle?

5) Multi-bridge. For certain studies, we may need to consider multiple ego vehicles (and thus multiple Autoware instances) connected to Carla. We understand each of them will need its own bridge. Is there any effort in this direction? What kind of development would be required to achieve this?

Thank you in advance for your answers.

Mario

**Comment by Robotics010** [comment]
Hey @MarioAlmelaUMH thank you for your interest in the bridge.

First of all the most of features and limitations are from the original carla's [ros-bridge](https://github.com/carla-simulator/ros-bridge). My `carla_autoware_bridge` is just adding a few missing things to ros-bridge in order to support Autoware Universe, so I can surely say about the `carla_autoware_bridge`, but I will not be so confident about the carla's [ros-bridge](https://github.com/carla-simulator/ros-bridge), because I don't know it fully. Nevertheless I'll try to answer your questions mixing the carla's ros-bridge and my carla_autoware_bridge. Also some of your questions were ahead of the [bridge 0.1 release](https://github.com/Robotics010/carla_autoware_bridge/tree/v0.1) docs, so I refer to the recent docs as well.

1. Work in progress status means that not all minimal initially planned features are released. [Currently](https://github.com/Robotics010/carla_autoware_bridge/tree/v0.1#current-limitations) (as for 0.1 version) supports only autoware galactic version, control module configuration and sensors calibration for a demo vehicle are absent. Control module doesn't have a proper configuration/calibration for the selected carla.tesla.model3 vehicle, so that longitudinal movement are not acceptable (breaking to strong) and lateral movement have more error than it should be (drives one wheel onto a curb on turns). Original carla's sensors from thier demo were adjusted but there still need to spend some effort to it. Also the bridge was not tested thoroughly in any application, so I would be expecting bugs in the bridge.
2. Currently I'm planning to investigate how to apply carla's synchronous mode with autoware and make some efforts to support up to humble (master) branch of Autoware. Next steps I was going to define later with the feedback from the bridge users. 
3. That should be definitely in the roadmap to check support scenarios execution within carla + autoware, but currently I can not say if it works already
4. Honestly I'm awating performance problems when using multiple lidars with the bridge, but it should be possible to mitigate this performance problem with synchronous mode (by not running in real-time). But it still need to be proved. The bridge should be able to support both multiple cameras/multiple lidars and custom sensor kits.
5. Currently it won't work if you try to spawn and use mutiple ego vehicles and it should be investigated to understand what kind of effort it will require. I would start from reading about support of multiple ego vehicles in CARLA simulator first and then would continue to think how to support it from the bridge side.

**Comment by Robotics010** [comment]
And a few questions to you @MarioAlmelaUMH if I may.

- Are you able, as a university official, communicate with a CARLA simulator and carla's ros-bridge developers? These developers are from some subdivision of Intel as I understand. I'm asking that because I was trying to ask questions about the carla's ros-bridge, I haven't got any answers and because of that making my own version of the bridge. It still would be better to have carla developers supporting autoware and making some efforts there.
- You were referring to some other useful carla-autoware candidates to select from. Can name a few of them? Because I know only the custom carla bridge ([CARLA Simulator + Autoware Universe Tutorials](https://github.com/orgs/autowarefoundation/discussions/2828)) from Hatem Darweesh, Nagoya University.

**Comment by MarioAlmelaUMH** [comment]
Hello, @Robotics010 
Regarding your questions:

- I'm still a rookie researcher, so I don't know how much influence would I have towards Carla's developers. As far as I know, my superiors already tried to contact them at the start of the project and Carla's team answered that they are not interested in developing an official bridge at the moment. On Autoware's side, I saw plans to develop bridges to many simulators like AWSIM and MORAI-Sim, and Carla was listed in some of those plans, however Autoware's team said that compatibility with Carla was difficult and I'm not sure when this bridge will be released. 
- The bridge you mention in your question, the one from Hatem Darweesh, is the one we are using right now. Besides it, we previously tested this old bridge (https://github.com/carla-simulator/carla-autoware), that uses a very early version of Carla and Autoware. We also tested bridges that work with other simulators or stacks, like Autoware's AWSIM bridge (https://tier4.github.io/AWSIM/GettingStarted/QuickStartDemo/), and Carla's Apollo bridge (https://github.com/guardstrikelab/carla_apollo_bridge).
